django/applications/catmaid/static/js/layers/pixi-image-block-layer.js

```diff
@@ -0,0 +1,395 @@
+        this._oldZoom = this._swapZoom;
+        this._oldZ    = this._swapZ;
+        this._renderIfReady();
+      }
+
+      if (typeof completionCallback !== 'undefined') {
+        if (loading && blocking) {
+          this._completionCallback = completionCallback;
+        } else {
+          this._completionCallback = null;
+          completionCallback();
+        }
+      }
+    }
+
+    _readBlock(zoomLevel, x, y, z) {
+      let blockCoord = CATMAID.tools.permute([x, y, z], this.recipDimPerm);
+
+      return this._blockCache.readBlock(zoomLevel, ...blockCoord)
+        .then(block => {
+          if (block) block = block.transpose(...this.dimPerm);
+
+          return block;
+        });
+    }
+
+    _sliceBlock(block, blockZ) {
+      var slice;
+      if (block && block.shape[2] > blockZ) {
+        slice = block.pick(null, null, blockZ);
+
+        if (slice.shape[0] < this.tileWidth ||
+            slice.shape[1] < this.tileHeight) {
+          let empty = this._makeEmptySlice();
+          var sub = empty.hi(slice.shape[0], slice.shape[1]);
+
+          for(let i=0; i<slice.shape[0]; ++i) {
+            for(let j=0; j<slice.shape[1]; ++j) {
+              empty.set(i,j, slice.get(i, j));
+            }
+          }
+        }
+      } else {
+        slice = this._makeEmptySlice();
+      }
+
+      return slice;
+    }
+
+    _makeEmptySlice() {
+      let empty = nj.zeros(
+          [this.tileWidth, this.tileHeight],
+          this.tileSource.dataType());
+      empty.selection.data.fill(this.fillValue);
+
+      return empty;
+    }
+
+    _dtypeWebGLParams(dtype) {
+      const gl = this._context.renderer.gl;
+      var format, type, internalFormat, jsArrayType;
+
+      switch (dtype) {
+        case 'uint8':
+          format = gl.LUMINANCE;
+          type = gl.UNSIGNED_BYTE;
+          internalFormat = gl.LUMINANCE;
+          jsArrayType = Uint8Array;
+          break;
+        // The default case can be hit when the layer is drawn before the
+        // image block source has fully loaded.
+        default:
+          CATMAID.warn(`Unknown data type for stack layer: ${dtype}, using uint32`);
+          /* falls through */
+        case 'uint32':
+          format = gl.RGBA;
+          type = gl.UNSIGNED_BYTE;
+          internalFormat = gl.RGBA;
+          jsArrayType = Uint8Array;
+          break;
+      }
+
+      return {format, type, internalFormat, jsArrayType};
+    }
+
+    _sliceToTexture(slice, pixiTex) {
+      let renderer = this._context.renderer;
+      let gl = renderer.gl;
+
+      let {format, type, internalFormat, jsArrayType} = this._dtypeWebGLParams(slice.dtype);
+      const glScaleMode = this._pixiInterpolationMode === PIXI.SCALE_MODES.LINEAR ?
+        gl.LINEAR : gl.NEAREST;
+
+      let baseTex = pixiTex.baseTexture;
+      let texture = baseTex._glTextures[renderer.CONTEXT_UID];
+      let newTex = false;
+      let width = slice.shape[1];
+      let height = slice.shape[0];
```

Comment:
I suppose, `shape[1]` is width and `shape[0]` and height because of the `transpose(1, 0)` call in line 193?